src/mongoc/cnv-mongoc-client-pool.c

```diff
@@ -160,64 +135,80 @@ cnv_mongoc_client_pool_pop (cnv_mongoc_client_pool_t *pool)
 
    ENTRY;
 
-   bson_return_val_if_fail(pool, NULL);
+   BSON_ASSERT(pool);
 
    mongoc_mutex_lock(&pool->mutex);
 
-   if (!((pooledClient = (_pooled_mongoc_client_t *) _mongoc_queue_pop_head(&pool->queue)))) {
-     // We don't want to create new client inside lock, too much contention unnecessarily
-     mongoc_mutex_unlock(&pool->mutex);
-
-     client = mongoc_client_new_from_uri(pool->uri);
+   if (!(pooledClient = (_pooled_mongoc_client_t *) _mongoc_queue_pop_head(&pool->queue))) {
+     client = mongoc_client_new_from_uri(pool->uri, pool->topology);
 #ifdef MONGOC_ENABLE_SSL
      if (pool->ssl_opts_set) {
        mongoc_client_set_ssl_opts (client, &pool->ssl_opts);
      }
 #endif
-   } else {     
-     mongoc_mutex_unlock(&pool->mutex);
+   } else {
      client = pooledClient->client;
      bson_free(pooledClient);
    }
 
+   mongoc_mutex_unlock(&pool->mutex);
```

Comment:
Notice there is additional argument here pool->topology, this is shared data and probably it requires locking. I'll check implementation to see how it's used